### PR TITLE
rgw: multisite enabled over multiple clusters

### PR DESCRIPTION
**Pull request by alimaredia**
Added '--cluster' to all necessary commands
ex: radosgw-admin, rados, ceph, made sure
necessary checks were in place so that clients
can be read with our without a cluster_name
preceeding them

Made master_client defined in the config for
radosgw-admin task

Signed-off-by: Ali Maredia <amaredia@redhat.com>

### CODE REVIEW

**Comment by oritwas** [comment]
I don't like duplicate parameters and the configuration already has the master zone client.
you can use get_config_master_client  from task/utils/rgw.py

**Comment by cbodley** [comment]
@alimaredia ping

File: qa/suites/rgw/multisite/all/multi_cluster.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
+meta:
+- desc: run two ceph clusters 
+roles:
+- - cluster1.mon.a
+  - cluster1.mon.b
+  - cluster1.osd.0
+  - cluster1.osd.1
+  - cluster1.osd.2
+  - cluster2.mon.c
+  - cluster1.client.0
```
Comment (oritwas) [inline qa/suites/rgw/multisite/all/multi_cluster.yaml]: it will be easier to read if the cluster weren't mixed also order by component

File: qa/suites/rgw/multisite/all/multi_cluster.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  - cluster2.osd.2
+  - cluster2.mgr.y
+tasks:
+- install:
+- ceph:
+    cluster: cluster1
+    conf:
+      client.0:
+        rgw region: jerry
+        rgw zone: seinfeld
```
Comment (oritwas) [inline qa/suites/rgw/multisite/all/multi_cluster.yaml]: great name :+1:

Comment (oritwas) [inline qa/suites/rgw/multisite/all/multi_cluster.yaml]: we will need to clean up the region at some point (not in this pr)

File: qa/tasks/radosgw_admin.py [hunk 3]
```diff
@@ -63,28 +62,30 @@ def remove_newlines(string):
...
-                   for id_ in teuthology.all_roles_of_type(ctx.cluster, 'client')]
-    if config is None:
-        config = all_clients
-    if isinstance(config, list):
-        config = dict.fromkeys(config)
-    clients = config.keys()
 
-    multi_region_run = rgw_utils.multi_region_enabled(ctx)
+    # regions and config found from rgw task
+    regions = ctx.rgw.regions
```
Comment (oritwas) [inline qa/tasks/radosgw_admin.py]: check to see there are regions

File: qa/tasks/radosgw_admin.py [hunk 4]
```diff
@@ -63,28 +62,30 @@ def remove_newlines(string):
...
+    config = ctx.rgw.config
```
Comment (oritwas) [inline qa/tasks/radosgw_admin.py]: check to see there is a config

File: qa/tasks/radosgw_agent.py [hunk 5]
```diff
@@ -18,8 +18,9 @@ def run_radosgw_agent(ctx, config):
     return_list = list()
     for (client, cconf) in config.items():
         # don't process entries that are not clients
-        if not client.startswith('client.'):
-            log.debug('key {data} does not start with \'client.\', moving on'.format(
+    	log.debug("client is %r", client)
```
Comment (oritwas) [inline qa/tasks/radosgw_agent.py]: extra space

File: qa/tasks/rgw.py [hunk 6]
```diff
@@ -62,19 +39,22 @@ def create_apache_dirs(ctx, config, on_client = None, except_client = None):
     for client in clients_to_create_as:
         if client == except_client:
             continue
+    	cluster_name, daemon_type, client_id = teuthology.split_role(client)
+    	# in case client doesn't have cluster_name attached to it
+    	client_with_cluster = cluster_name + '.' + daemon_type + '.' + client_id
```
Comment (oritwas) [inline qa/tasks/rgw.py]: What is the cluster name is empty?

File: qa/tasks/rgw.py [hunk 7]
```diff
@@ -134,6 +114,10 @@ def ship_apache_configs(ctx, config, role_endpoints, on_client = None,
     for client in clients_to_create_as:
         if client == except_client:
             continue
+    	cluster_name, daemon_type, client_id = teuthology.split_role(client)
+    	# in case client doesn't have cluster_name attached to it
+    	client_with_id = daemon_type + '.' + client_id
+    	client_with_cluster = cluster_name + '.' + client_with_id
```
Comment (oritwas) [inline qa/tasks/rgw.py]: what if cluster_name is empty?

File: qa/tasks/rgw.py [hunk 8]
```diff
@@ -260,18 +245,24 @@ def start_rgw(ctx, config, on_client = None, except_client = None):
     """
     Start rgw on remote sites.
     """
-    log.info('Starting rgw...')
+    log.info('ALI ADDED Starting rgw...')
```
Comment (oritwas) [inline qa/tasks/rgw.py]: don't forget to remove this :)

File: qa/tasks/rgw.py [hunk 9]
```diff
@@ -260,18 +245,24 @@ def start_rgw(ctx, config, on_client = None, except_client = None):
...
+        log.debug('client %r', clients_to_run)
     testdir = teuthology.get_testdir(ctx)
     for client in clients_to_run:
         if client == except_client:
             continue
         (remote,) = ctx.cluster.only(client).remotes.iterkeys()
+        cluster_name, daemon_type, client_id = teuthology.split_role(client)
+        client_with_id = daemon_type + '.' + client_id
+        # in case client doesn't have cluster_name attached to it
+        client_with_cluster = cluster_name + '.' + client_with_id
```
Comment (oritwas) [inline qa/tasks/rgw.py]: see previous comment

File: qa/tasks/rgw.py [hunk 10]
```diff
@@ -134,6 +114,10 @@ def ship_apache_configs(ctx, config, role_endpoints, on_client = None,
     for client in clients_to_create_as:
         if client == except_client:
             continue
+    	cluster_name, daemon_type, client_id = teuthology.split_role(client)
+    	# in case client doesn't have cluster_name attached to it
+    	client_with_id = daemon_type + '.' + client_id
```
Comment (oritwas) [inline qa/tasks/rgw.py]: I think the comment is wrong , you need if for when you provide the cluster separately

Comment (oritwas) [inline qa/tasks/rgw.py]: Also please check that daemon_type and client_id are not empty

File: qa/tasks/rgw.py [hunk 11]
```diff
@@ -177,32 +161,33 @@ def ship_apache_configs(ctx, config, role_endpoints, on_client = None,
             fcgi_config = f.read()
         with file(src, 'rb') as f:
             conf = f.read() + fcgi_config
+            log.debug('ALI conf is %r', conf)
```
Comment (oritwas) [inline qa/tasks/rgw.py]: please remove

File: qa/tasks/rgw.py [hunk 12]
```diff
@@ -217,21 +202,21 @@ def ship_apache_configs(ctx, config, role_endpoints, on_client = None,
...
             remote=remote,
-            path='{tdir}/apache/htdocs.{client}/rgw.fcgi'.format(
+            path='{tdir}/apache/htdocs.{client_with_cluster}/rgw.fcgi'.format(
                 tdir=testdir,
-                client=client),
+                client_with_cluster=client_with_cluster),
             data="""#!/bin/sh
 ulimit -c unlimited
-exec radosgw -f -n {client} -k /etc/ceph/ceph.{client}.keyring {rgw_options}
+exec radosgw -f -n {client_with_id} --cluster {cluster_name} -k /etc/ceph/{client_with_cluster}.keyring {rgw_options}
```
Comment (oritwas) [inline qa/tasks/rgw.py]: will it be simpler to use --id instead of -n?

File: qa/tasks/rgw.py [hunk 13]
```diff
@@ -260,18 +245,24 @@ def start_rgw(ctx, config, on_client = None, except_client = None):
...
         if client == except_client:
             continue
         (remote,) = ctx.cluster.only(client).remotes.iterkeys()
+        cluster_name, daemon_type, client_id = teuthology.split_role(client)
+        client_with_id = daemon_type + '.' + client_id
+        # in case client doesn't have cluster_name attached to it
+        client_with_cluster = cluster_name + '.' + client_with_id
         zone = rgw_utils.zone_for_client(ctx, client)
         log.debug('zone %s', zone)
+
```
Comment (oritwas) [inline qa/tasks/rgw.py]: nit empty line

File: qa/tasks/rgw.py [hunk 14]
```diff
@@ -393,6 +386,8 @@ def start_apache(ctx, config, on_client = None, except_client = None):
     if on_client is None:
         clients_to_run = config.keys()
     for client in clients_to_run:
+        cluster_name, daemon_type, client_id = teuthology.split_role(client)
+        client_with_cluster = cluster_name + '.' + daemon_type + '.' + client_id
```
Comment (oritwas) [inline qa/tasks/rgw.py]: please check for empty cluster_name and daemon_type

File: qa/tasks/rgw.py [hunk 15]
```diff
@@ -636,6 +541,7 @@ def configure_users_for_client(ctx, config, client, everywhere=False):
             continue
 
         for client_name in clients_to_create_as:
+            cluster_name, daemon_type, client_id = teuthology.split_role(client_name)
```
Comment (oritwas) [inline qa/tasks/rgw.py]: not used, please remove

File: qa/tasks/rgw.py [hunk 16]
```diff
@@ -676,6 +582,7 @@ def configure_users(ctx, config,  everywhere=False):
         if everywhere:
             clients_to_create_as = config.keys()
         for client_name in clients_to_create_as:
+            cluster_name, daemon_type, client_id = teuthology.split_role(client_name)
```
Comment (oritwas) [inline qa/tasks/rgw.py]: not used, please remove

File: qa/tasks/rgw.py [hunk 17]
```diff
@@ -738,6 +647,7 @@ def configure_multisite_regions_and_zones(ctx, config, regions, role_endpoints,
     log.debug('regions are %r', regions)
     log.debug('role_endpoints = %r', role_endpoints)
     log.debug('realm is %r', realm)
+
```
Comment (oritwas) [inline qa/tasks/rgw.py]: nit empty line

File: qa/suites/rgw/multisite/all/multi_cluster.yaml [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+        rgw region root pool: .rgw.region.0
```
Comment (oritwas) [inline qa/suites/rgw/multisite/all/multi_cluster.yaml]: We should use `rgw zonegroup root pool ` and not `rgw region root pool`
Also what about` rgw period root pool` and `rgw realm root pool`?

File: qa/tasks/rgw.py [hunk 19]
```diff
@@ -349,6 +341,7 @@ def start_rgw(ctx, config, on_client = None, except_client = None):
 
         ctx.daemons.add_daemon(
             remote, 'rgw', client,
+            cluster=cluster_name,
```
Comment (oritwas) [inline qa/tasks/rgw.py]: make sure it works if cluster_name is empty

File: qa/tasks/rgw.py [hunk 20]
```diff
@@ -870,6 +781,7 @@ def configure_regions_and_zones(ctx, config, regions, role_endpoints, realm):
     log.debug('regions are %r', regions)
     log.debug('role_endpoints = %r', role_endpoints)
     log.debug('realm is %r', realm)
+
```
Comment (oritwas) [inline qa/tasks/rgw.py]: nit

File: qa/tasks/rgw.py [hunk 21]
```diff
@@ -893,14 +805,15 @@ def configure_regions_and_zones(ctx, config, regions, role_endpoints, realm):
     fill_in_endpoints(region_info, role_zones, role_endpoints)
 
     # clear out the old defaults
-    first_mon = teuthology.get_first_mon(ctx, config)
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
Comment (oritwas) [inline qa/tasks/rgw.py]: see previous comments

File: qa/tasks/rgw.py [hunk 22]
```diff
@@ -1029,26 +943,35 @@ def pull_configuration(ctx, config, regions, role_endpoints, realm, master_clien
...
                      user_info['system_key']['secret_key']],
                      check_status=True)
 
             (zonegroup, zone, zone_info, zone_user_info) = role_zones[client]
             zone_json = json.dumps(dict(zone_info.items() + zone_user_info.items()))
-            log.debug("zone info is: %r"), zone_json
+            log.debug("zone info is: %r", zone_json)
             rgwadmin(ctx, client,
-                     cmd=['zone', 'set', '--rgw-zonegroup', zonegroup,
+                     cmd=['zone', 'set', '--default',
```
Comment (oritwas) [inline qa/tasks/rgw.py]: why did you change this. this will set and add the zone to the zonegroup

File: qa/tasks/rgw.py [hunk 23]
```diff
@@ -1029,26 +943,35 @@ def pull_configuration(ctx, config, regions, role_endpoints, realm, master_clien
...
                           '--rgw-zone', zone],
                      stdin=StringIO(zone_json),
                      check_status=True)
 
             rgwadmin(ctx, client,
+                     cmd=['zonegroup', 'add', '--rgw-zonegroup', zonegroup, '--rgw-zone', zone],
```
Comment (oritwas) [inline qa/tasks/rgw.py]: no need if you don't change the set command

File: qa/tasks/rgw.py [hunk 24]
```diff
@@ -1256,6 +1179,7 @@ def task(ctx, config):
         lambda: create_nonregion_pools(
             ctx=ctx, config=config, regions=regions),
         ]
+    log.debug('Nonregion pools created')
```
Comment (oritwas) [inline qa/tasks/rgw.py]: missing tab

File: qa/tasks/s3tests.py [hunk 25]
```diff
@@ -29,7 +29,8 @@ def extract_sync_client_data(ctx, client_name):
     """
     return_region_name = None
     return_dict = None
-    client = ctx.ceph['ceph'].conf.get(client_name, None)
+    cluster_name, daemon_type, client_id = teuthology.split_role(client_name)
```
Comment (oritwas) [inline qa/tasks/s3tests.py]: what if cluster_name is empty?

File: qa/tasks/s3tests.py [hunk 26]
```diff
@@ -199,19 +200,22 @@ def create_users(ctx, config):
         for section, user in users.iteritems():
             _config_user(s3tests_conf, section, '{user}.{client}'.format(user=user, client=client))
             log.debug('Creating user {user} on {host}'.format(user=s3tests_conf[section]['user_id'], host=client))
+            cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
Comment (oritwas) [inline qa/tasks/s3tests.py]: missing checks

File: qa/tasks/util/rgw.py [hunk 27]
```diff
@@ -20,14 +19,17 @@ def rgwadmin(ctx, client, cmd, stdin=StringIO(), check_status=False,
              format='json'):
     log.info('rgwadmin: {client} : {cmd}'.format(client=client,cmd=cmd))
     testdir = teuthology.get_testdir(ctx)
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
Comment (oritwas) [inline qa/tasks/util/rgw.py]: see previous comment

File: qa/tasks/util/rgw.py [hunk 28]
```diff
@@ -70,8 +72,10 @@ def get_user_successful_ops(out, user):
     return get_user_summary(out, user)['total']['successful_ops']
 
 def get_zone_host_and_port(ctx, client, zone):
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
Comment (oritwas) [inline qa/tasks/util/rgw.py]: see previous comment

File: qa/tasks/util/rgw.py [hunk 29]
```diff
@@ -85,8 +89,10 @@ def get_zone_host_and_port(ctx, client, zone):
     assert False, 'no endpoint for zone {zone} found'.format(zone=zone)
 
 def get_master_zone(ctx, client):
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
Comment (oritwas) [inline qa/tasks/util/rgw.py]: see previous comment

File: qa/tasks/util/rgw.py [hunk 30]
```diff
@@ -116,27 +122,29 @@ def get_master_client(ctx, clients):
...
-                                 'zone', 'get', '--rgw-zone', zone])
+                            cmd=['zone', 'get', '--rgw-zone', zone])
     system_key = zone_info['system_key']
     return system_key['access_key'], system_key['secret_key']
 
 def zone_for_client(ctx, client):
-    ceph_config = ctx.ceph['ceph'].conf.get('global', {})
-    ceph_config.update(ctx.ceph['ceph'].conf.get('client', {}))
-    ceph_config.update(ctx.ceph['ceph'].conf.get(client, {}))
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
Comment (oritwas) [inline qa/tasks/util/rgw.py]: see previous comment

File: qa/tasks/util/rgw.py [hunk 31]
```diff
@@ -116,27 +122,29 @@ def get_master_client(ctx, clients):
...
+    ceph_config = ctx.ceph[cluster_name].conf.get('global', {})
+    ceph_config.update(ctx.ceph[cluster_name].conf.get('client', {}))
+    ceph_config.update(ctx.ceph[cluster_name].conf.get(client, {}))
     return ceph_config.get('rgw zone')
 
 def region_for_client(ctx, client):
-    ceph_config = ctx.ceph['ceph'].conf.get('global', {})
-    ceph_config.update(ctx.ceph['ceph'].conf.get('client', {}))
-    ceph_config.update(ctx.ceph['ceph'].conf.get(client, {}))
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
Comment (oritwas) [inline qa/tasks/util/rgw.py]: see previous comment

File: qa/tasks/util/rgw.py [hunk 32]
```diff
@@ -116,27 +122,29 @@ def get_master_client(ctx, clients):
...
+    ceph_config = ctx.ceph[cluster_name].conf.get('global', {})
+    ceph_config.update(ctx.ceph[cluster_name].conf.get('client', {}))
+    ceph_config.update(ctx.ceph[cluster_name].conf.get(client, {}))
     return ceph_config.get('rgw region')
 
 def radosgw_data_log_window(ctx, client):
-    ceph_config = ctx.ceph['ceph'].conf.get('global', {})
-    ceph_config.update(ctx.ceph['ceph'].conf.get('client', {}))
-    ceph_config.update(ctx.ceph['ceph'].conf.get(client, {}))
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
Comment (oritwas) [inline qa/tasks/util/rgw.py]: see previous comment

File: qa/tasks/util/rgw.py [hunk 33]
```diff
@@ -181,3 +189,116 @@ def get_sync_agent(ctx, source):
...
     return None, None
+
+def extract_zone_info(ctx, client, client_config):
+    """
+    Get zone information.
+    :param client: dictionary of client information
+    :param client_config: dictionary of client configuration information
+    :returns: zone extracted from client and client_config information
+    """
+    cluster_name, daemon_type, client_id = teuthology.split_role(client)
```
Comment (oritwas) [inline qa/tasks/util/rgw.py]: see previous comment

File: qa/tasks/rgw.py (hunk 6 above)
Comment (alimaredia) [inline qa/tasks/rgw.py]: It defaults to "ceph"

File: qa/tasks/rgw.py (hunk 8 above)
Comment (alimaredia) [inline qa/tasks/rgw.py]: Yes i plan to do a cleanup of the entire PR. I just add that in so I can easily ctrl+f important parts of the logs.

File: qa/tasks/s3tests.py (hunk 26 above)
Comment (alimaredia) [inline qa/tasks/s3tests.py]: what's the check I'm missing? Did you mean to ask what if the cluster_name is empty?

Comment (oritwas) [inline qa/tasks/s3tests.py]: yes , but you explained it is defaults to ceph so that is ok

**Comment by alimaredia** [comment]
@oritwas @cbodley I removed the multisite directory from qa/suites/rgw. I want to get these changes merged now because they don't, and shouldn't hinder any functionality.

I cleaned up the PR per Orit's last run through the code, but one last runthrough from both of you would be nice. Below are all the runs of teuthology to show that these changes don't break everything:

rgw/verify:
http://pulpito.ceph.com/amaredia-2017-04-05_14:46:26-rgw:verify-wip-ali-testing---basic-smithi/

rgw/multifs (one run for each frontend):
http://pulpito.ceph.com/amaredia-2017-04-03_17:47:50-rgw:multifs-master---basic-mira/
http://pulpito.ceph.com/amaredia-2017-04-04_04:11:13-rgw:multifs-master---basic-vps/

rgw/singleton:
http://pulpito.ceph.com/amaredia-2017-04-04_19:11:15-rgw:singleton-wip-cbodley-testing---basic-mira/

File: qa/tasks/rgw.py (hunk 22 above)
Comment (alimaredia) [inline qa/tasks/rgw.py]: great questions. @cbodley do you remember? I remember there was an issue and this fix solved it, but I can't remember more than that.

File: qa/tasks/rgw.py (hunk 23 above)
Comment (alimaredia) [inline qa/tasks/rgw.py]: @cbodley I'm pretty sure this change had to do with the last one for reference.

File: qa/tasks/rgw.py (hunk 12 above)
Comment (alimaredia) [inline qa/tasks/rgw.py]: im not sure, does it make a difference? @cbodley have an opinion?

File: qa/tasks/rgw.py (hunk 24 above)
Comment (alimaredia) [inline qa/tasks/rgw.py]: I think it's fine because it's aligned with subtasks above it and multisite right below it?

File: qa/tasks/rgw.py (hunk 22 above)
Comment (cbodley) [inline qa/tasks/rgw.py]: @oritwas it doesn't look like 'zone set' calls `RGWZoneGroup::add_zone()`, so we had to add the `zonegroup add` below

**Comment by cbodley** [comment]
@alimaredia for those teuthology runs you linked, were you using the `cluster_name.` syntax in any of the yaml?

**Comment by alimaredia** [comment]
@cbodley nope, whatever yaml is existing in the rgw suite now was used

File: qa/tasks/rgw.py (hunk 22 above)
Comment (oritwas) [inline qa/tasks/rgw.py]: I see ...
I will fix that ...

**Comment by alimaredia** [comment]
the branch has been rebased and updated, and these thrash tests were run: http://pulpito.ceph.com/amaredia-2017-04-13_18:45:09-rgw:thrash-wip-ali-testing---basic-smithi/

**Comment by cbodley** [comment]
thanks @alimaredia, looks good to me

@oritwas can you take a look?
